idp_task1.cpp

```diff
@@ -2,15 +2,43 @@
 #include "socket_exception.hpp"
 #include <iostream>
 #include <string.h>
+#include <vector>
+#include <thread>
 
 static const int BUFFER_MAXLEN = 100;
 static const int LISTENQ = 16;
 static const int SERVER_PORT = 8877;
 static const char* ECHO_STR_PREFIX = "echo: ";
 static const int ECHO_STR_LEN = strlen(ECHO_STR_PREFIX);
 
+void listeningThread(Socket_sptr client_socket)
+{
+    int n = 0;
+    int maxlen = 100;
+    char buffer[maxlen];
+
+    char echo[maxlen + ECHO_STR_LEN];
+    strcpy(echo, ECHO_STR_PREFIX);
+    memset(buffer, 0, maxlen);
+
+    while ((n = client_socket->Recv(buffer, maxlen, 0)) > 0)
```

Comment:
And exception thrown by socket calls will not be caught and will lead to application crash.
You'll need to process all exceptions in each thread function using `try ... catch`. If you really want to propagate exception processing to another thread, the C++11 feature `exception_ptr` will give you a chance.